### DISCUSSION TITLE
Possible to pass extension/args to basic middleware functions?

### DISCUSSION THREAD

**Discussion by ra0x3** [tokio-rs/axum#1829] [Q&A] [answered]
- I'm referencing the following example from the Axum docs [here](https://docs.rs/axum/latest/axum/middleware/index.html#passing-state-from-middleware-to-handlers)
- I have a piece of code that basically mirrors that example ☝🏽 
- However, in my code, (referencing the snippet above) in my `authorize_current_user` function, I'd like to authorize the user based on the value of some config.
  - This config is used on other route handlers in my code as an `Extension`
- I'm pretty much having trouble figuring out if it's even possible to add an argument to the `auth` function (in the snippet above).
- Ideally I'm trying to implement the following:

```rust
use axum::{
    Router,
    http::{Request, StatusCode},
    routing::get,
    response::{IntoResponse, Response},
    middleware::{self, Next},
    extract::Extension,
};

#[derive(Clone)]
struct CurrentUser { /* ... */ }

async fn auth<B>(mut req: Request<B>, next: Next<B>, config: Config) -> Result<Response, StatusCode> {
    let auth_header = req.headers()
        .get(http::header::AUTHORIZATION)
        .and_then(|header| header.to_str().ok());

    let auth_header = if let Some(auth_header) = auth_header {
        auth_header
    } else {
        return Err(StatusCode::UNAUTHORIZED);
    };

    if let Some(current_user) = authorize_current_user(auth_header, config).await {
        // insert the current user into a request extension so the handler can
        // extract it
        req.extensions_mut().insert(current_user);
        Ok(next.run(req).await)
    } else {
        Err(StatusCode::UNAUTHORIZED)
    }
}

async fn authorize_current_user(auth_token: &str, config: Config) -> Option<CurrentUser> {
    // authorize the user using `auth_token`, but based on some value in `Config`
}

async fn handler(
    // extract the current user, set by the middleware
    Extension(current_user): Extension<CurrentUser>,
) {
    // ...
}

let app = Router::new()
    .route("/", get(handler))
    .route_layer(middleware::from_fn(auth));
```

**Comment by davidpdrsn** [comment, marked as answer]
Yes. See https://docs.rs/axum/latest/axum/middleware/index.html#accessing-state-in-axummiddlewarefrom_fn

**Reply by ra0x3** [reply]
@davidpdrsn Quick follow up.

- The implementation in the link you shared works perfectly 🙏🏽 
- However, one other question I had (not sure if this should be asked in a separate question) is if I use your suggested implementation ☝🏽 , how am I supposed to add an `Extension` onto my route handler?
- Using your suggested implementation, I still plan to insert an `Extension` into my request (`req.extensions_mut().insert(current_user)`) before calling `self.inner.call(req)`.
- However, not really sure if I can just add an `Extension` in my route handler in order to access that data, [similar to how this works with simple state-less middlware functions](https://docs.rs/axum/latest/axum/middleware/index.html#passing-state-from-middleware-to-handlers)
- As I'm trying it now, getting some gnarly errors related to `the trait `Handler<_, _, _>` is not implemented for fn item...`

**Reply by davidpdrsn** [reply]
> However, one other question I had (not sure if this should be asked in a separate question) is if I use your suggested implementation ☝🏽 , how am I supposed to add an Extension onto my route handler?

`Extension` is also an extractor so you can just add it as a function argument like any other extractor.

> As I'm trying it now, getting some gnarly errors related to the trait Handler<_, _, _> is not implemented for fn item...

Use https://docs.rs/axum/latest/axum/attr.debug_handler.html for those.

**Reply by ra0x3** [reply]
@davidpdrsn Hopefully this is my last question 😅 
- But it seems I'm getting a `Missing request extension: Extension of type ___` response
- I have my stateful middleware implementation [here](https://github.com/FuelLabs/fuel-indexer/pull/630/files#diff-6d7a0f70d732b8cd90686608882fa1ffee31bed12549df78dd6ffc561995e480R48) (confirming the proper codepath is hit)
- I'm injecting the middleware in question [here](https://github.com/FuelLabs/fuel-indexer/pull/630/files#diff-6d7a0f70d732b8cd90686608882fa1ffee31bed12549df78dd6ffc561995e480R85) 
- I'm adding that middleware layer onto my router [here](https://github.com/FuelLabs/fuel-indexer/pull/630/files#diff-d302ec562581a109feb66c372bafcd6bed3c523a9b1733a1ff856673c0b45110R144)
- Finally, I have my `Extension` in question in my handler [here](https://github.com/FuelLabs/fuel-indexer/pull/630/files#diff-2e1f3139bd6582f4d7866145625a9e02c27e47edfd5f10be8c945a2436ef8fe9R132)
- Everything appears to be hooked up from what I can see (going by the docs)
- I'm pretty certain I'm overlooking something. Just not sure what that something might be, from reading the docs

🤔 

(Apologies for not having a more ready minimum reproducible example)

Edit: as expected, was a me issue :)
